utils/hare-status

```diff
@@ -167,31 +148,91 @@ def processes(cns: Consul, consul_util: ConsulUtil,
         parts = x['Key'].split('/', 5)
         if len(parts) == 5:
             proc_fidks.append(fid_key(parts[-1]))
+    node_health = consul_util.get_node_health_status(node_name)
+    process_states = get_node_process_states(cns, node_name)
+    process_list = get_node_processes(cns, node_name)
     return [
         Process(
-            name=proc_id2name(cns, node_name, k),
+            name=proc_id2name(process_list, k),
             fid=Fid(0x7200000000000001, k),
-            ep=(
-                kv_value_as_str(
-                    cns,
-                    f'm0conf/nodes/{node_name}/processes/{k}/endpoint'
-                ) or '**ERROR**'),
-            status=process_status(consul_util, node_name,
+            ep=(proc_id2endpoint(process_list, k) or '**ERROR**'),
+            status=process_status(node_health, process_states,
                                   Fid(0x7200000000000001, k)))
         for k in sorted(proc_fidks)
     ]
 
 
-def process_status(consul_util: ConsulUtil, node_name, fid: Fid) -> str:
-    if (consul_util.get_node_health_status(node_name) != 'passing'):
+def process_status(node_health: str, proc_states: Dict[str, str],
+                   fid: Fid) -> str:
+    if node_health != 'passing':
         return 'offline'
-    status = consul_util.get_process_status(fid, node_name)
-    if status.proc_status == 'M0_CONF_HA_PROCESS_STARTED':
-        return 'started'
-    elif status.proc_status in ('Unknown', 'UNKNOWN'):
-        return 'unknown'
+    if str(fid) in proc_states:
+        return proc_states[str(fid)]
     else:
-        return 'offline'
+        return 'unknown'
+
+
+def get_node_process_states(cns: Consul, node_name: str) -> Dict[str, str]:
+    key = f'{node_name}/processes/'
+    proc_list = kv_item(cns, key, recurse=True)
+    process_states: Dict[str, str] = {}
+    for proc in proc_list:
+        key_split = proc['Key'].split('/')
+        if len(key_split) != 3:
+            continue
+        fid = key_split[-1]
+        val = json.loads(proc['Value'])
+        if val['state'] == 'M0_CONF_HA_PROCESS_STARTED':
+            state = 'started'
+        elif val['state'] in ('Unknown', 'UNKNOWN'):
+            state = 'unknown'
+        else:
+            state = 'offline'
+        process_states[fid] = state
+    return process_states
+
+
+def proc_id2endpoint(process_list: List[Dict[str, Any]],
```

Comment:
Here we are iterating on each consul kv, another option is to use, `consul_util.catalog.get_services()`, check `get_service_data_by_name(service_name)` this will give all the details for a given service type on a node including its status and endpoint. So we don't have to fetch endpoint individually for every process.
For process status we still need to fetch it from consul kv but for endpoint we can use this.

Reply:
`get_service_data_by_name(service_name)` This function indeed returns details for that **particular service name** but on **all nodes**.

That would help for multiple processes with the same name like ios and s3, only. For rest, we need to make individual consul kv calls. 
Had it would return all the services on a particular node, would be actually helpful.
This is what I understand, please correct me.

> Here we are iterating on each consul kv,

Yes, correct. But we are parsing the data iteratively from the recurse key of 'm0conf/nodes/{node_name}/processes/',  which is only fetched once.
Also, we could optimize it and store all fid-> endpoint keys mapping in a dictionary once and use it.

Reply:
Okay, we can also get a node level details,
```
`[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-1473 /]# curl http://127.0.0.1:8500/v1/catalog/node/cortx-data-headless-svc-ssc-vm-g2-rhev4-1473
{"Node":{"ID":"f85ebb3b-5eba-1879-7976-b899c975fb17","Node":"cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Address":"172.16.84.98","Datacenter":"dc1","TaggedAddresses":{"lan":"172.16.84.98","lan_ipv4":"172.16.84.98","wan":"172.16.84.98","wan_ipv4":"172.16.84.98"},"Meta":{"consul-network-segment":""},"CreateIndex":8299,"ModifyIndex":8306},"Services":{"259":{"ID":"259","Service":"hax","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":2001,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8310,"ModifyIndex":8310},"262":{"ID":"262","Service":"ios","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":3001,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8315,"ModifyIndex":8315},"275":{"ID":"275","Service":"ios","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":3002,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8316,"ModifyIndex":8316},"288":{"ID":"288","Service":"confd","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":3003,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8313,"ModifyIndex":8313},"291":{"ID":"291","Service":"s3service","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":4001,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8317,"ModifyIndex":8317},"294":{"ID":"294","Service":"s3service","Tags":[],"Address":"inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-1473","Meta":null,"Port":4002,"Weights":{"Passing":1,"Warning":1},"EnableTagOverride":false,"Proxy":{"MeshGateway":{},"Expose":{}},"Connect":{},"CreateIndex":8318,"ModifyIndex":8318}}}[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-1473 /]#
```
this will give you endpoints for all the services on a given node. Idea it to avoid pattern matching because that incurs multilevel string comparisons and in worst case the string we are trying to find can be at the bottom of the list. If we can have some kind of an associative array it may help.

But, on the second thought, let's take this as an improvement and complete this first version.

Reply:
@mssawant  do we want to take it to main branch as well?